### PR TITLE
Feature/babyjub verify

### PR DESCRIPTION
**Pull request by arnaucube**
- Refactors from old spec to current spec:
  - removes old spec related
  - imports last rootcommits.sol
  - migrates from Root to ID State
- Adds Poseidon
- Adds BabyJubJub signature verification

Poseidon and BabyJubJub verification compatible with [go-iden3-crypto](https://github.com/iden3/go-iden3-crypto) implementations

### CODE REVIEW

File: contracts/lib/Iden3Helpers.sol [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+ * @dev Define interface Mimc7 contract
+ */
+contract MimcUnit {
+  function MiMCpe7(uint256,uint256) public pure returns(uint256) {}
+}
+
+/**
+ * @dev Iden3 helper functions
+ */
+contract Iden3Helpers {
```
Comment (ed255) [inline contracts/lib/Iden3Helpers.sol]: The functions in this file seem to be outated.  We no longer use mimc7.  We don't use claims that authorize ethereum keys.  Claims are 8 entries.

File: contracts/lib/Memory.sol [hunk 2]
```diff
@@ -9,130 +9,200 @@ library Memory {
...
        }
        return Cursor(ptr,ptr+self.length);
     }
-    
+
     /**
-    * @dev reads 32 bytes from cursor, no eof checks 
-    * @return the value
+    * @dev reads 32 bytes from cursor, no eof checks
+    * @return b the value
```
Comment (ed255) [inline contracts/lib/Memory.sol]: typo

File: contracts/lib/MiMc7.sol [hunk 3]
```diff
@@ -0,0 +1,46 @@
+pragma solidity ^0.5.0;
```
Comment (ed255) [inline contracts/lib/MiMc7.sol]: Where did this come from?  Mimc7 is outdated.

File: contracts/test/Iden3HelpersTest.sol [hunk 4]
```diff
@@ -0,0 +1,113 @@
+pragma solidity ^0.5.0;
```
Comment (ed255) [inline contracts/test/Iden3HelpersTest.sol]: Where did this file come from?

File: test/Iden3Helpers.test.js [hunk 5]
```diff
@@ -0,0 +1,217 @@
+/* global artifacts */
```
Comment (ed255) [inline test/Iden3Helpers.test.js]: This is outdated (mimc7), where did this come from?

File: test/mimc7.test.js [hunk 6]
```diff
@@ -0,0 +1,61 @@
+/* global artifacts */
```
Comment (ed255) [inline test/mimc7.test.js]: where did this file come from?

**Comment by ed255** [comment]
While I was checking the diff there were many "new files" that had what seems to be old code (mimc7, claims with 4 entries, etc.).  I assume this code was moved from somewhere to the new files, but it was hard to follow what's new and what's old but moved in this PR.  So before accepting I would like to review this again with @arnaucube on my side.

File: contracts/lib/Iden3Helpers.sol (hunk 1 above)
Comment (arnaucube) [reply contracts/lib/Iden3Helpers.sol]: Yes, Iden3Helpers are from the old spec, is not used currently, probably will be deleted at the end of the current phase (not yet, as now is in current active development of the branch)

File: contracts/lib/MiMc7.sol (hunk 3 above)
Comment (arnaucube) [reply contracts/lib/MiMc7.sol]: This repo was outdated, and Mimc7 was in the private/internal smart contracts with most of the rest of the contracts to publish identity roots/states.
On the first commit of this PR was added as general import of the last version of the smart contracts (that was in the private repo)
This is the commit that imports it together with the rest of contracts that were outdated in this repo: https://github.com/iden3/contracts/commit/27e36bc52540a6ca0a2c26fd06e8c253b99d7995

File: test/Iden3Helpers.test.js (hunk 5 above)
Comment (arnaucube) [reply test/Iden3Helpers.test.js]: See comments above

File: test/mimc7.test.js (hunk 6 above)
Comment (arnaucube) [reply test/mimc7.test.js]: See comments above

File: contracts/test/Iden3HelpersTest.sol (hunk 4 above)
Comment (arnaucube) [reply contracts/test/Iden3HelpersTest.sol]: See comments above

**Comment by arnaucube** [comment]
Closing this PR without merging. Working in another branch in the ECC arithmetics and the BabyJubJub signature verification
